### ISSUE TITLE
Build under Python 3.12 fails on pip upgrade step

### ISSUE THREAD

**Issue by brettcannon** [readthedocs/readthedocs.org#10832]
## Details

* Read the Docs project URL: https://readthedocs.org/projects/microvenv/
* Build URL (if applicable): [https://readthedocs.org/projects/{your_project_slug}/builds/{build_id}/](https://readthedocs.org/projects/microvenv/builds/22259147/)
* Read the Docs username (if applicable): https://readthedocs.org/profiles/{your_rtd_username}/

## Expected Result

Build to succeed.

## Actual Result

```
python -m pip install --upgrade --no-cache-dir pip setuptools
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/__main__.py", line 29, in <module>
    from pip._internal.cli.main import main as _main
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_internal/cli/main.py", line 9, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_internal/cli/autocompletion.py", line 10, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_internal/cli/main_parser.py", line 8, in <module>
    from pip._internal.cli import cmdoptions
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_internal/cli/cmdoptions.py", line 23, in <module>
    from pip._internal.cli.parser import ConfigOptionParser
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_internal/cli/parser.py", line 12, in <module>
    from pip._internal.configuration import Configuration, ConfigurationError
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_internal/configuration.py", line 21, in <module>
    from pip._internal.exceptions import (
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_internal/exceptions.py", line 7, in <module>
    from pip._vendor.pkg_resources import Distribution
  File "/home/docs/checkouts/readthedocs.org/user_builds/microvenv/envs/57/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2164, in <module>
    register_finder(pkgutil.ImpImporter, find_on_path)
                    ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
```

To work around this I have pinned to Python 3.11 in my `.readthedocs.yml` file.

**Comment by stsewd** [comment]
Linking to build https://readthedocs.org/projects/microvenv/builds/22259147/.

Found a couple of issues in Cpython/pip/setuptools

- https://github.com/python/cpython/issues/109779
- https://github.com/python/cpython/issues/108541
- https://github.com/pypa/pip/issues/12179

Looks like maybe related to the pre-installed version of pip/setuptools

https://github.com/readthedocs/readthedocs.org/blob/0154576f170b8de7ea7c9f35d9b74d4573ff200b/scripts/compile_version_upload_s3.sh#L111-L113

**Comment by stsewd** [comment]
I was able to reproduce this locally.

```console
$ pyenv install 3.12.0
$ pyenv virtualenv 3.12.0 test-3.12 
$ pyenv activate test-3.12
$ pip install 'pip==21.2.4'
$ pip install --upgrade --no-cache-dir pip
Traceback (most recent call last):
  File "/home/stsewd/.pyenv/versions/test-3.12/bin/pip", line 5, in <module>
    from pip._internal.cli.main import main
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_internal/cli/main.py", line 9, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_internal/cli/autocompletion.py", line 10, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_internal/cli/main_parser.py", line 8, in <module>
    from pip._internal.cli import cmdoptions
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_internal/cli/cmdoptions.py", line 23, in <module>
    from pip._internal.cli.parser import ConfigOptionParser
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_internal/cli/parser.py", line 12, in <module>
    from pip._internal.configuration import Configuration, ConfigurationError
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_internal/configuration.py", line 21, in <module>
    from pip._internal.exceptions import (
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_internal/exceptions.py", line 7, in <module>
    from pip._vendor.pkg_resources import Distribution
  File "/home/stsewd/.pyenv/versions/3.12.0/envs/test-3.12/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2164, in <module>
    register_finder(pkgutil.ImpImporter, find_on_path)
                    ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
```

**Comment by stsewd** [comment]
We have changed the default of `python: "3"` to 3.11.x till we fix the problem with 3.12.
